bee-crypto/src/ternary/bigint/i384/mod.rs

```diff
@@ -0,0 +1,893 @@
+                }
+            }
+        }
+    };
+}
+
+impl I384<BigEndian, U8Repr> {
+    /// Applies not on all bytes of the I384.
+    pub fn not_inplace(&mut self) {
+        for digit in &mut self.inner[..] {
+            *digit = !*digit;
+        }
+    }
+}
+
+impl_default!((I384<BigEndian, U8Repr>), LEN_IN_U8);
+
+impl Eq for I384<BigEndian, U8Repr> {}
+
+impl From<I384<BigEndian, U32Repr>> for I384<BigEndian, U8Repr> {
+    fn from(value: I384<BigEndian, U32Repr>) -> Self {
+        let mut i384_u8 = Self::zero();
+        byteorder::BigEndian::write_u32_into(&value.inner, &mut i384_u8.inner);
+        i384_u8
+    }
+}
+
+impl PartialEq for I384<BigEndian, U8Repr> {
+    fn eq(&self, other: &Self) -> bool {
+        let mut are_equal = true;
+        for (a, b) in self.inner.iter().zip(other.inner.iter()) {
+            are_equal &= a == b
+        }
+        are_equal
+    }
+}
+
+impl PartialOrd for I384<BigEndian, U8Repr> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        use Ordering::*;
+
+        let self_iter = self.inner.iter();
+        let other_iter = other.inner.iter();
+
+        let mut zipped_iter = self_iter.zip(other_iter);
+
+        // The most significant u8 (MSU8) has to be handled separately.
+        //
+        // If the most significant bit of both numbers is set, then the comparison operators have to be reversed.
+        //
+        // Note that this is only relevant to the comparison operators between the less significant u8 if the two MSU8s
+        // are equal. If they are not equal, then an early return will be triggered.
+
+        const NEGBIT: u8 = 0x80;
+        const UMAX: u8 = std::u8::MAX;
+        let numbers_negative = match zipped_iter.next() {
+            // Case 1: both numbers are negative, s is less.
+            Some((s @ NEGBIT..=UMAX, o @ NEGBIT..=UMAX)) if s > o => return Some(Greater),
+
+            // Case 2: both numbers are negative, s is greater.
+            Some((s @ NEGBIT..=UMAX, o @ NEGBIT..=UMAX)) if s < o => return Some(Less),
+
+            // Case 3: both numbers are negative, but equal.
+            Some((NEGBIT..=UMAX, NEGBIT..=UMAX)) => true,
+
+            // Case 4: only s is negative.
+            Some((NEGBIT..=UMAX, _)) => return Some(Less),
+
+            // Case 5: only o is negative.
+            Some((_, NEGBIT..=UMAX)) => return Some(Greater),
+
+            // Case 6: both are positive, s is greater.
+            Some((s, o)) if s > o => return Some(Greater),
+
+            // Case 7: both are positive, s is less.
+            Some((s, o)) if s < o => return Some(Less),
+
+            // Fallthrough case; only happens if s == o.
+            Some(_) => false,
+
+            // The array inside `I384` always has a length larger zero, so the first element is guaranteed to exist.
+            None => unreachable!(),
+        };
+
+        // Create two separate loops as to avoid repeatedly checking `numbers_negative`.
```

Comment:
Not much point in doing this, compilers have been doing this sort of iteration-invariant lifting optimisation since the 90s. This probably just makes it harder for the compiler to reason about the code.